File: src/jaslib_src/jaslib/ui/web/vocab/vocab_kanji_list_renderer.py
from __future__ import annotations

from typing import TYPE_CHECKING

from jaslib.viewmodels.kanji_list import sentence_kanji_list_viewmodel
from jaspythonutils.sysutils import ex_str

if TYPE_CHECKING:
    from jaslib.note.vocabulary.vocabnote import VocabNote


def render_kanji_list_from_kanji(kanjis: list[str]) -> str:
    if not kanjis:
        return ""

    viewmodel = sentence_kanji_list_viewmodel.create(kanjis)

    return f"""
<div id="kanji_list" class="page_section">
    <div class="page_section_title">kanji</div>
{ex_str.newline.join(f'''
    <div class="kanji_item {" ".join(kanji.kanji.get_meta_tags())}">
        <div class="kanji_main">
            <span class="kanji_kanji clipboard">{kanji.question()}</span>
            <span class="kanji_answer">{kanji.answer()}</span>
            <span class="kanji_readings">{kanji.readings()}</span>
        </div>
        <div class="kanji_mnemonic">{kanji.mnemonic()}</div>
    </div>
''' for kanji in viewmodel.kanji_list)}
</div>
        """

def render_vocab_kanji_list(vocab: VocabNote) -> str:
    return render_kanji_list_from_kanji(vocab.kanji.extract_main_form_kanji())

File: src/src_dotnet/JAStudio.Core/UI/Web/Vocab/VocabKanjiListRenderer.cs
using JAStudio.Core.ViewModels.KanjiList;
using System.Collections.Generic;
using System.Linq;
using JAStudio.Core.Note.Vocabulary;

namespace JAStudio.Core.UI.Web.Vocab;

public class VocabKanjiListRenderer
{
    readonly SentenceKanjiListViewModel _sentenceKanjiListViewModel;
   internal VocabKanjiListRenderer(SentenceKanjiListViewModel sentenceKanjiListViewModel) => _sentenceKanjiListViewModel = sentenceKanjiListViewModel;

    public string RenderKanjiListFromKanji(List<string> kanjis)
    {
        if (kanjis.Count == 0)
            return "";

        var viewmodel = _sentenceKanjiListViewModel.Create(kanjis);

        var kanjiItems = viewmodel.KanjiList.Select(kanji => $$$"""
    <div class="kanji_item {{{string.Join(" ", kanji.Kanji.GetMetaTags())}}}">
        <div class="kanji_main">
            <span class="kanji_kanji clipboard">{{{kanji.Question()}}}</span>
            <span class="kanji_answer">{{{kanji.Answer()}}}</span>
            <span class="kanji_readings">{{{kanji.Readings()}}}</span>
        </div>
        <div class="kanji_mnemonic">{{{kanji.Mnemonic()}}}</div>
    </div>
""");

        return $"""
            <div id="kanji_list" class="page_section">
                <div class="page_section_title">kanji</div>
            {string.Join("\n", kanjiItems)}
            </div>
            """;
    }

    public string RenderVocabKanjiList(VocabNote vocab)
    {
        return RenderKanjiListFromKanji(vocab.Kanji.ExtractMainFormKanji());
    }
}
